src/session_manager.rs

```diff
@@ -0,0 +1,1180 @@
+//! includes.
+//!
+//! # Adding new sessions
+//! The `SessionManager` directs it to [`Login`](struct@crate::login::Login) as soon as the user
+//! wants to add a new session within the function `add_new_session`. Login returns control to the
+//! session manager if either the new session is ready, logging into the new session is aborted by
+//! the user, or it is noticed that the phone number is already logged in (QR code session
+//! duplicates can be detected earliest if the phone number was determined using `fetch_me`).
+//!
+//! # Adding existing sessions
+//! The `SessionManager` analyzes the individual database directories in the Telegrand data
+//! directory to see which sessions can be logged in directly using
+//! [`SessionManager::add_existing_session()`]. To do this, it checks whether a specially created
+//! file called `session_info.json` exists in the respective database directory, in which it is
+//! stored whether the session was logged in.
+//!
+//! # Tearing down sessions
+//! The `SessionManager` uses a two-step process for this.
+//! The first step is to rename `session_info.json` within the database directory. The second step
+//! is to logout from td and to remove the database directory.
+//! See (see [`two_step_database_teardown()`]) for details.
+//! Tearing down sessions happens in different places: When the login is canceled, When the QR code
+//! is canceled, When a logged in session is logged out, and for all sessions that are not logged
+//! in when the application exits.
+//!
+//! # Migration strategy
+//! To migrate the previous database directory `db0`, this directory is searched at the first start
+//! and the `session_info.json` is written into it.
+//! There is one case where the migration strategy might not work. This case is if `db0` was a test
+//! account. To my knowledge there is no way to find this out before. Therefore we always assume
+//! that the session was running on a production Data Center.
+//!
+//! # Remembering recently used sessions
+//! In order to remember the order in which the user selected the sessions, the `SessionManager`
+//! maintains a file called `recently_used_sessions.json` in the data directory.
+
+use anyhow::anyhow;
+use futures::{Future, TryFutureExt, TryStreamExt};
+use glib::clone;
+use gtk::{gio, glib, prelude::*, subclass::prelude::*, CompositeTemplate};
+use serde::{Deserialize, Serialize};
+use std::borrow::Borrow;
+use std::io;
+use std::time::{SystemTime, UNIX_EPOCH};
+use tdgrand::enums::{self, AuthorizationState, Update};
+use tdgrand::functions;
+use tdgrand::types::{self, UpdateAuthorizationState};
+use tokio::fs::{self, File};
+use tokio::io::AsyncReadExt;
+use tokio_stream::wrappers::ReadDirStream;
+
+use crate::utils::{
+    data_dir, do_async, send_tdlib_parameters, session_info_path, write_session_info_to_file,
+};
+use crate::Session;
+use crate::APPLICATION_OPTS;
+use crate::RUNTIME;
+
+const LEAGACY_DB_NAME: &str = "db0";
```

Comment:
Do we really need this? I'd prefer not keeping references to old behavior since the app is unreleased anyway. I guess we can detect this db by just iterating in the data dir?

Reply:
Yes, this is indeed not needed.
In this move, no more `session_info.json` will be written to the database directory. Because I had completely overlooked that it is possible to distinguish between test accounts and production accounts by looking if either a `td.binlog` or `td_test.binlog` exists.
So now this information does not need to be written to an extra file.